scriptworker_client/src/scriptworker_client/aio.py

```diff
@@ -81,6 +89,163 @@
+            replacement dictionary values. e.g. ``{"path": {"find": "replace"}, ...}``
+
+    """
+
+    def __init__(
+        self,
+        coro,
+        lockfile_map,
+        args=None,
+        kwargs=None,
+        lockfile_kwargs=None,
+        retry_async_kwargs=None,
+        use_retry_async=False,
+    ):
+        """Initialize the ``LockfileFuture``.
+
+        Args:
+            coro (typing.Awaitable): the coroutine to run.
+            args (list, optional): the args to pass to the coroutine. Defaults to ``[]``
+            kwargs (dict, optional): the kwargs to pass to the coroutine. Defaults to ``{}``
+            retry_async_kwargs (dict, optional): the kwargs to pass to ``retry_async``. Defaults to ``{}``.
+            use_retry_async (bool, optional): whether to use ``retry_async``. Defaults to ``False``.
+            lockfile_kwargs (dict, optional): the kwargs to pass to ``lockfile``. Defaults to ``{}``.
+            lockfile_map (dict): a mapping between lockfile path keys to arg/kwarg
+                replacement dictionary values. e.g. ``{"path": {"find": "replace"}, ...}``
+
+        """
+        self.coro = coro
+        self.args = args or ()
+        self.kwargs = kwargs or {}
+        self.lockfile_map = lockfile_map
+        self.lockfile_kwargs = lockfile_kwargs or {}
+        self.retry_async_kwargs = retry_async_kwargs or {}
+        self.use_retry_async = use_retry_async
+
+    def replace_args(self, obj, repl_dict):
+        """Do string sprintf replacement for all strings in `obj`.
+
+        Args:
+            obj (object): an object to do string replacement in. Non-
+                strings, lists, and dicts will be passed through unchanged.
+            repl_dict (dict): the replacement dictionary. If a string has,
+                e.g. ``%(foo)s`` in it, then a ``repl_dict`` of
+                ``{"foo": "bar"}`` will replace the ``%(foo)s`` with ``bar``.
+
+        Returns:
+            obj, with string replacement if applicable.
+
+        """
+        if isinstance(obj, str):
+            return obj % repl_dict
+        elif isinstance(obj, (list, tuple)):
+            return [self.replace_args(item, repl_dict) for item in obj]
+        elif isinstance(obj, dict):
+            return {key: self.replace_args(val, repl_dict) for key, val in obj.items()}
+        else:
+            return obj
+
+    async def run_with_lockfile(self):
+        """Run the coro with the lockfile.
+
+        Get a lockfile, update ``self.args`` and ``self.kwargs`` with the
+        corresponding ``self.lockfile_map`` value, then await ``self.coro``
+        (wrapped with ``retry_async`` if ``self.use_retry_async``).
+
+        """
+        async with lockfile(
+            self.lockfile_map.keys(), **self.lockfile_kwargs
+        ) as lockfile_path:
+            args = self.replace_args(self.args, self.lockfile_map[lockfile_path])
+            kwargs = self.replace_args(self.kwargs, self.lockfile_map[lockfile_path])
```

Comment:
Doing replacement in all of the args/kwargs seems a bit excessive considering the only purpose in real usage (as I understand it) is to pass the username to the coro. Have you considered passing that as a separate arg or kwarg instead? It's hard for me to envision whether or not that's viable/better, but I wanted to raise it at least.

Reply:
If the coro is, for example, `run_command`, then my args might be something like `["sudo", "su", "%(user)s", "-c", "bash blah blah"]`. By allowing the recursive replacement, I can do this directly. Without it, I have to write a wrapper coro that takes the user arg and other info, to call `run_command`.